File: src/main/java/no/nav/syfo/service/HistorikkService.java
package no.nav.syfo.service;

import lombok.extern.slf4j.Slf4j;
import no.nav.syfo.consumer.ws.PersonConsumer;
import no.nav.syfo.domain.rest.*;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestClientException;

import javax.inject.Inject;
import java.util.ArrayList;
import java.util.List;

import static java.util.stream.Collectors.toList;
import static no.nav.syfo.util.MapUtil.mapListe;

@Slf4j
@Service
public class HistorikkService {

    private final MotebehovService motebehovService;
    private final VeilederOppgaverService veilederOppgaverService;
    private final PersonConsumer personConsumer;

    public static final String HAR_SVART_PAA_MOTEBEHOV = " har svart på møtebehov";
    public static final String MOTEBEHOVET_BLE_LEST_AV = "Møtebehovet ble lest av ";

    @Inject
    public HistorikkService(
            MotebehovService motebehovService,
            VeilederOppgaverService veilederOppgaverService,
            PersonConsumer personConsumer
    ) {
        this.veilederOppgaverService = veilederOppgaverService;
        this.motebehovService = motebehovService;
        this.personConsumer = personConsumer;
    }

    public List<Historikk> hentHistorikkListe(final Fnr arbeidstakerFnr) {
        String fnr = arbeidstakerFnr.getFnr();

        List<Historikk> historikkListe = hentOpprettetMotebehov(arbeidstakerFnr);
        historikkListe.addAll(hentLesteMotebehovHistorikk(fnr));

        return historikkListe;
    }

    private List<Historikk> hentOpprettetMotebehov(Fnr arbeidstakerFnr) {
        List<Motebehov> motebehovListe = motebehovService.hentMotebehovListe(arbeidstakerFnr);

        return mapListe(
                motebehovListe,
                motebehov -> new Historikk()
                        .opprettetAv(motebehov.opprettetAv)
                        .tekst(personConsumer.hentNavnFraAktoerId(motebehov.opprettetAv) + HAR_SVART_PAA_MOTEBEHOV)
                        .tidspunkt(motebehov.opprettetDato)
        );
    }

    private List<Historikk> hentLesteMotebehovHistorikk(String sykmeldtFnr) {
        try {
            return veilederOppgaverService.get(sykmeldtFnr)
                    .stream()
                    .filter(veilederOppgave -> veilederOppgave.type.equals("MOTEBEHOV_MOTTATT") && veilederOppgave.status.equals("FERDIG"))
                    .map(veilederOppgave -> new Historikk()
                            .tekst(MOTEBEHOVET_BLE_LEST_AV + veilederOppgave.sistEndretAv)
                            .tidspunkt(veilederOppgave.getSistEndretAsLocalDateTime())
                    )
                    .collect(toList());
        } catch (RestClientException e) {
            log.error("Klarte ikke hente ut varselhistorikk på leste møtebehov");
            return new ArrayList<>();
        }
    }

}

File: src/main/kotlin/no/nav/syfo/motebehov/historikk/HistorikkService.kt
package no.nav.syfo.motebehov.historikk

import no.nav.syfo.consumer.pdl.PdlConsumer
import no.nav.syfo.consumer.pdl.fullName
import no.nav.syfo.motebehov.Motebehov
import no.nav.syfo.motebehov.MotebehovService
import org.springframework.stereotype.Service
import java.util.stream.Collectors
import javax.inject.Inject

@Service
class HistorikkService @Inject constructor(
    private val motebehovService: MotebehovService,
    private val pdlConsumer: PdlConsumer
) {
    fun hentHistorikkListe(arbeidstakerFnr: String): List<Historikk> {
        val motebehovListe = motebehovService.hentMotebehovListe(arbeidstakerFnr)

        val historikkListe = hentOpprettetMotebehov(motebehovListe)
        historikkListe.addAll(hentBehandlendeMotebehovHistorikk(motebehovListe))
        return historikkListe
    }

    private fun hentOpprettetMotebehov(motebehovListe: List<Motebehov>): MutableList<Historikk> {
        return motebehovListe.map {
            Historikk(
                opprettetAv = it.opprettetAv,
                tekst = "${getNameOfCreatedBy(it)}$HAR_SVART_PAA_MOTEBEHOV",
                tidspunkt = it.opprettetDato
            )
        }.toMutableList()
    }

    private fun getNameOfCreatedBy(motebehov: Motebehov): String {
        val createdByFnr = pdlConsumer.fnr(motebehov.opprettetAv)
        return pdlConsumer.person(createdByFnr)?.fullName() ?: ""
    }

    private fun hentBehandlendeMotebehovHistorikk(motebehovListe: List<Motebehov>): List<Historikk> {
        return motebehovListe
            .stream()
            .filter { motebehov -> !motebehov.behandletVeilederIdent.isNullOrEmpty() && motebehov.behandletTidspunkt != null }
            .map { motebehov ->
                Historikk(
                    tekst = MOTEBEHOVET_BLE_LEST_AV + motebehov.behandletVeilederIdent,
                    tidspunkt = motebehov.behandletTidspunkt!!
                )
            }
            .collect(Collectors.toList())
    }

    companion object {
        const val HAR_SVART_PAA_MOTEBEHOV = " har svart på møtebehov"
        const val MOTEBEHOVET_BLE_LEST_AV = "Møtebehovet ble behandlet av "
    }
}
